src/slots.rs

```diff
@@ -0,0 +1,110 @@
+
+use log::{debug, log_enabled, Level};
+use prometheus_exporter::prometheus::IntCounterVec;
+use solana_client::{client_error::ClientError, rpc_client::RpcClient};
+use solana_sdk::epoch_info::EpochInfo;
+use std::collections::BTreeMap;
+
+/// The monitor of skipped and validated slots per validator with minimal internal state.
+pub struct SkippedSlotsMonitor<'a> {
+    /// Shared Solana RPC client.
+    client: &'a RpcClient,
+    /// Prometheus gauge.
+    leader_slots: &'a IntCounterVec,
+    /// The last observed epoch number.
+    epoch_number: u64,
+    /// The last observed slot index.
+    slot_index: u64,
+    /// The slot leader schedule for the last observed epoch.
+    slot_leaders: BTreeMap<usize, String>,
+    /// `true` iff `SkippedSlotMonitor::export_skipped_slots` already ran.
+    already_ran: bool,
+}
+
+impl<'a> SkippedSlotsMonitor<'a> {
+    /// Constructs a monitor given `client`.
+    pub fn new(client: &'a RpcClient, leader_slots: &'a IntCounterVec) -> Self {
+        Self {
+            client,
+            leader_slots,
+            epoch_number: 0,
+            slot_index: 0,
+            slot_leaders: Default::default(),
+            already_ran: false,
+        }
+    }
+
+    /// Exports the skipped slot statistics given `epoch_info`.
+    pub fn export_skipped_slots(&mut self, epoch_info: &EpochInfo) -> Result<(), ClientError> {
+        let first_slot = epoch_info.absolute_slot - epoch_info.slot_index;
+
+        if self.epoch_number != epoch_info.epoch {
+            self.slot_leaders = self.get_slot_leaders(None)?;
+            self.epoch_number = epoch_info.epoch;
+            self.slot_index = epoch_info.slot_index;
+            debug!("SkippedSlotsMonitor state updated");
+        } else if self.slot_index == epoch_info.slot_index {
+            debug!("At the slot index");
+            return Ok(());
+        }
+
+        let (abs_range_start, range_start) = if self.already_ran {
+            // Start from the last seen slot if already ran before.
+            (first_slot + self.slot_index, self.slot_index)
+        } else {
+            // Start from the first slot in the current epoch if running for the first time.
+            self.already_ran = true;
+            (first_slot, 0)
+        };
+        let range_end = epoch_info.slot_index;
+        let abs_range_end = first_slot + range_end;
+
+        let confirmed_blocks = self
+            .client
+            .get_confirmed_blocks(abs_range_start, Some(abs_range_end))?;
+        debug!(
+            "Confirmed blocks from {} to {}: {:?}",
+            abs_range_start, abs_range_end, confirmed_blocks
+        );
+        let mut feed = self.leader_slots.local();
+        for slot_in_epoch in range_start..range_end {
+            let leader = &self.slot_leaders[&(slot_in_epoch as usize)];
+            let absolute_slot = first_slot + slot_in_epoch;
+            let status = if confirmed_blocks.binary_search(&absolute_slot).is_ok() {
```

Comment:
`status` here can only be one of two values. This probably be defined concretely at the type-level using an enum?